perun/view/scatter_plot/factory.py

```diff
@@ -0,0 +1,156 @@
+            uid_models(list)
+    """
+    for uid in uids:
+        # Slice only the plotted uid from the data table
+        uid_slice = resources[resources.uid == uid]
+        if uid_slice.size == 0 or uid_slice.shape[0] <= 1:
+            # plotting one point does not work (it has no real usage anyway), fix later
+            continue
+        # Filter models for the given uid
+        uid_models = list(filter(lambda m, u=uid: m['uid'] in u, models))
+        yield uid_slice, uid_models
+
+
+def slice_models_by_interval(models):
+    """ Splits the models list into slices with different x axis intervals.
+
+    Arguments:
+        models(list of dict): the list of models to split
+
+    Returns:
+        generator: stream of models slices (list)
+    """
+    # Sort the models by intervals first, to yield them in order
+    models = sorted(models, key=itemgetter('x_interval_start', 'x_interval_end'))
+    # Separate the models into groups according to intervals
+    intervals = defaultdict(list)
+    for model in models:
+        intervals[(model['x_interval_start'], model['x_interval_end'])].append(model)
+    # Yield the list of models with the same interval
+    for interval_models in intervals.items():
+        yield interval_models[1]
+
+
+def generate_plot_data_slices(profile):
+    """ Generates data slices for plotting resources and models. The resources are split by unique
+        uids, models are sliced into parts by uid and interval.
+
+    Arguments:
+        profile(dict): loaded perun profile
+
+    Returns:
+        generator: generator: resources and models slices of unique uid as pair
+            data_slice(pandas.DataFrame)
+            uid_models(list)
+    """
+    # Get resources for scatter plot points and models for curves
+    resource_table = converters.resources_to_pandas_dataframe(profile)
+    models = [m[1] for m in list(query.all_models_of(profile))]
+    # Get unique uids from profile, each uid (and optionally interval) will have separate graph
+    uids = query.unique_resource_values_of(profile, 'uid')
+
+    # Process each uid data
+    for uid_slice, uid_models in slice_resources_by_uid(resource_table, models, uids):
+        # Slice the uid models according to different intervals (each interval is plotted
+        # separately as it improves readability)
+        if uid_models:
+            for interval_models in slice_models_by_interval(uid_models):
+                yield uid_slice, interval_models
+        else:
+            # There are no models to plot
+            yield uid_slice, []
+
+
+def draw_models(graph, models):
+    """ Add models renderers to the graph.
+
+    Arguments:
+        graph(charts.Graph): the scatter plot without models
+        models(list): list of models to plot
+
+    Returns:
+        charts.Graph: the modified graph with model curves renderers
+    """
+    # Get unique colors for the model curves
+    colour_palette = palettes.viridis(len(models))
```

Comment:
Why not use colour generator from bokeh_helpers?

Reply:
This would require to transform the 'models' into pandas dataframe - right now the 'models' is a simple list of model values. The transformation will be solved in the full version eventually.

Reply:
Great, note to myself we should fix/change this in bokeh_helpers (after discussion of the remodelling though). Probably would be nice to have two swatches: (1) e.g. viridian (for dataframes---number would be obtained by unique keys), (2) e.g. inferno (for lists---i.e. the number would be obtained by len()).